libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -621,58 +621,38 @@ void ContractLevelChecker::checkBaseABICompatibility(ContractDefinition const& _
 
 void ContractLevelChecker::checkAmbiguousOverrides(ContractDefinition const& _contract) const
 {
-	vector<FunctionDefinition const*> contractFuncs = _contract.definedFunctions();
-
-	auto const resolvedBases = resolveDirectBaseContracts(_contract);
-
-	FunctionMultiSet inheritedFuncs = inheritedFunctions(&_contract);;
+	// Fetch non overridden base functions and sort them by signature.
+	FunctionMultiSet nonOverriddenBaseFunctionSet = convertContainer<FunctionMultiSet>(
+		nonOverriddenBaseFunctions(&_contract)
+	);
+	// Remove all functions that match the signature of a function in the current contract.
+	nonOverriddenBaseFunctionSet -= _contract.definedFunctions();
 
-	// Check the sets of the most-inherited functions
-	for (auto it = inheritedFuncs.cbegin(); it != inheritedFuncs.cend(); it = inheritedFuncs.upper_bound(*it))
+	// Walk through the set of functions signature by signature.
+	for (auto it = nonOverriddenBaseFunctionSet.cbegin(); it != nonOverriddenBaseFunctionSet.cend();)
 	{
-		auto [begin,end] = inheritedFuncs.equal_range(*it);
-
-		// Only one function
-		if (next(begin) == end)
-			continue;
+		auto nextSignature = nonOverriddenBaseFunctionSet.upper_bound(*it);
 
-		// Not an overridable function
-		if ((*it)->isConstructor())
+		// Only one function with this signature - skip it.
+		if (next(it) == nextSignature)
+			++it;
+		else
 		{
-			for (begin++; begin != end; begin++)
-				solAssert((*begin)->isConstructor(), "All functions in range expected to be constructors!");
-			continue;
-		}
+			// We have multiple non-overridden functions with the same signature in the bases.
+			auto const& name = (*it)->name();
 
-		// Function has been explicitly overridden
-		if (contains_if(
-			contractFuncs,
-			[&] (FunctionDefinition const* _f) {
-				return hasEqualNameAndParameters(*_f, **it);
-			}
-		))
-			continue;
-
-		set<FunctionDefinition const*> ambiguousFunctions;
-		SecondarySourceLocation ssl;
+			SecondarySourceLocation ssl;
+			for (;it != nextSignature; ++it)
+				ssl.append("Definition here: ", (*it)->location());
```

Comment:
Is there a test for when this loop runs > 1?

Reply:
It *always* runs more than once :-).